wrench/src/wrench.rs

```diff
@@ -0,0 +1,361 @@
+
+struct Notifier {
+    window_proxy: WindowProxy,
+}
+
+impl Notifier {
+    fn new(window_proxy: WindowProxy) -> Notifier {
+        Notifier {
+            window_proxy: window_proxy,
+        }
+    }
+}
+
+impl RenderNotifier for Notifier {
+    fn new_frame_ready(&mut self) {
+        self.window_proxy.wakeup_event_loop();
+    }
+
+    fn new_scroll_frame_ready(&mut self, _composite_needed: bool) {
+        self.window_proxy.wakeup_event_loop();
+    }
+
+    fn pipeline_size_changed(&mut self,
+                             _: PipelineId,
+                             _: Option<Size2D<f32>>) {
+    }
+}
+
+#[cfg(target_os = "windows")]
+pub fn layout_simple_ascii(face: NativeFontHandle, text: &str, size: Au) -> (Vec<u16>, Vec<f32>) {
+    let system_fc = dwrote::FontCollection::system();
+    let font = system_fc.get_font_from_descriptor(&(face as dwrote::FontDescriptor)).unwrap();
+    let face = font.create_font_face();
+
+    let chars: Vec<u32> = text.chars().map(|c| c as u32).collect();
+    let indices = face.get_glyph_indices(&chars);
+    let glyph_metrics = face.get_design_glyph_metrics(&indices, false);
+
+    let device_pixel_ratio: f32 = 1.0;
+    let em_size = size.to_f32_px() / 16.;
+    let design_units_per_pixel = face.metrics().designUnitsPerEm as f32 / 16. as f32;
+    let scaled_design_units_to_pixels = (em_size * device_pixel_ratio) / design_units_per_pixel;
+
+    let advances = glyph_metrics.iter().map(|m| m.advanceWidth as f32 * scaled_design_units_to_pixels).collect();
+
+    (indices, advances)
+}
+
+#[cfg(not(target_os = "windows"))]
+pub fn layout_simple_ascii(face: NativeFontHandle, text: &str, size: Au) -> (Vec<u16>, Vec<f32>) {
+    panic!("Can't layout simple ascii on this platform");
+}
+
+pub trait WrenchThing {
+    fn next_frame(&mut self);
+    fn prev_frame(&mut self);
+    fn do_frame(&mut self, &mut Wrench) -> u32;
+}
+
+pub struct Wrench {
+    pub window: glutin::Window,
+    pub window_size: Size2D<u32>,
+    pub device_pixel_ratio: f32,
+
+    pub renderer: webrender::renderer::Renderer,
+    pub sender: RenderApiSender,
+    pub api: RenderApi,
+
+    pub image_map: HashMap<PathBuf, (ImageKey, Size2D<f32>)>,
+
+    pub builder: Option<DisplayListBuilder>,
+    pub aux_builder: Option<AuxiliaryListsBuilder>,
+
+    // internal housekeeping
+    pub root_pipeline_id: PipelineId,
+    pub next_scroll_layer_id: usize,
+    pub frame_count: u32,
+
+    pub gl_renderer: String,
+    pub gl_version: String,
+}
+
+impl Wrench {
+    pub fn new(shader_override_path: Option<PathBuf>,
+               dp_ratio: Option<f32>,
+               subpixel_aa: bool,
+               record: bool,
+               debug: bool)
+           -> Wrench
+    {
+        // First create our GL window
+        let window = glutin::WindowBuilder::new()
+            .with_gl(glutin::GlRequest::Specific(glutin::Api::OpenGl, (3, 2)))
+            .build().unwrap();
+
+        unsafe {
+            window.make_current().ok();
```

Comment:
I think we should be using `unwrap()` here, because `ok()` will not crash, and our behavior would be undefined later on